### PR TITLE
Adding cacheable networkExecutionState to Executor

### PR DESCRIPTION
**Pull request by gcatron**
Summary:
This adds a NetworkExecutionState class to the executor, it is similar to the ExecutionState object except that it is initialized once and re-used for multiple executions.
This will reduce the overhead in the Executor and allows for further optimizations such as pinned memory in contexts, and P2P support. 
Documentation:

Test Plan:
Ensure unit tests pass

### CODE REVIEW

**Comment by gcatron** [comment]
@jsubag

**Comment by tracelogfb** [comment]
How does synchronization work on the cached state? It seems like we just round robin across object in the pool and wrap around. What's preventing from a thread from using an object that's still in use by another one?

It also seems like it's not a good idea for the ThreadPoolExecutor to block, right? That could easily lead to a dead lock situation if we ware not careful.

**Comment by yinghai** [comment]
Should we review this or this is still wip?

**Comment by gcatron** [comment]
> Should we review this or this is still wip?

Still WIP for now, I need to stitch in inputs/outputs from the request context. Also since we're going that route I don't think there's a need to keep executionState and NetworkExecutionState around so I'll fold them in. Should be ready later today.

File: tests/unittests/ThreadPoolExecutorTest.cpp [hunk 1]
```diff
@@ -696,7 +698,7 @@ TEST_F(ThreadPoolExecutorTest, ConcurrentSingleNode) {
   constexpr RunIdentifierTy baseTestRunId = 10;
   constexpr DeviceIDTy testDeviceId = 111;
   constexpr unsigned deviceManagerThreads = 3;
-  unsigned numConcurrentRuns = 1000;
+  unsigned numConcurrentRuns = 10;
```
Comment (yinghai) [inline tests/unittests/ThreadPoolExecutorTest.cpp]: Why this change?

File: include/glow/Runtime/Executor/NetworkExecutionState.h [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  bool initialized_{false};
+
+private:
+  /// The run identifier for this execution of a DAG.
+  RunIdentifierTy runId_;
+  /// The callback that should be called when execution is done.
+  ResultCBTy cb_;
+
+  /// Map of all buffers allocated for intermediate contexts.
+  std::unordered_map<Placeholder *, void *> buffers_;
```
Comment (yinghai) [inline include/glow/Runtime/Executor/NetworkExecutionState.h]: Can we group these members by things that are static (things we can determine when we create the state pool except for the tensors that we need to bind), and things are dynamic (mainly bookkeeping of the dag job status).

File: include/glow/Runtime/Executor/NetworkExecutionState.h [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  /// Thread Executor used for this run.
+  // ThreadExecutor *executor_;
+};
+
+class NetworkExecutionStatePool {
+public:
+  NetworkExecutionState *getNextNetworkExecutionState() {
+    unsigned current = currentState_;
+    currentState_ = (currentState_ + 1) % states_.size();
+    return states_[current].get();
```
Comment (yinghai) [inline include/glow/Runtime/Executor/NetworkExecutionState.h]: This is not thread safe. Two threads can read the same currentState_. Even with currentState_.fetch_and_add() it is still not safe. Here you can do what we did for the NNPI EnvPool, using a stack guarded by mutex. Or since we have access to boost now, we can use boost::lock_free::stack<NetExecutionState>.

File: lib/Runtime/HostManager/HostManager.cpp [hunk 4]
```diff
@@ -228,6 +228,11 @@ Error HostManager::addNetwork(std::unique_ptr<Module> module,
     return err;
   }
 
+  // Create pool of cachedExecutionStates.
+  for (auto &node : nodeList) {
```
Comment (yinghai) [inline lib/Runtime/HostManager/HostManager.cpp]: This potentially wastes quite a bit memory. Let's keep it this way and see whether the trade-off is worth it.

Comment (yinghai) [reply lib/Runtime/HostManager/HostManager.cpp]: Actually do we need to set default as number of inflight requests? I'd think it should be number of workers that will feed up the backend? Say 2?

File: include/glow/Runtime/Executor/NetworkExecutionState.h (hunk 3 above)
Comment (gcatron) [reply include/glow/Runtime/Executor/NetworkExecutionState.h]: I'm hoping to avoid a mutex since this is in the inference path, If I use fetch_add(), and move the modulus to line 142, It should be thread safe, states_.size() should not change after initialization. I can refactor create pool to make that more explicit.

Comment (yinghai) [reply include/glow/Runtime/Executor/NetworkExecutionState.h]: Modulus will wrap around.

Comment (gcatron) [reply include/glow/Runtime/Executor/NetworkExecutionState.h]: That's fine, since the size of the pool == maxActiveRequests. there won't be overlap.

Comment (yinghai) [reply include/glow/Runtime/Executor/NetworkExecutionState.h]: It's not safe. As I mentioned, having pool size equal to maxActiveRequests may not be a good idea. And relying on this implicitly may potentially lead to hard to debug issues. Mutex is actually not that bad and we have access to lockfree stack also. Such scheme has been proven to be efficient in large scale services.

File: tests/unittests/ThreadPoolExecutorTest.cpp (hunk 1 above)
Comment (gcatron) [reply tests/unittests/ThreadPoolExecutorTest.cpp]: set it to 10 for testing, changing it to 100. 1000 doesn't add much and really slows down testing particularly in debug.

Comment (yinghai) [reply tests/unittests/ThreadPoolExecutorTest.cpp]: Cool. We do want the system to be over subscribed to flush out some race issue. 100 seems to be good.

File: lib/Runtime/HostManager/HostManager.cpp (hunk 4 above)
Comment (gcatron) [reply lib/Runtime/HostManager/HostManager.cpp]: I think it needs to be maxActiveRequests for now. Since executor->run() modifies the networkExecutionState. say maxActive = 10 and workers = 2. If you have 10 requests, the first 2 would be enqueued and acted on immediately so we need at least 2. The next 8 would be dispatched to the executor, which would put them on the queue for the Device. Those 8 would need their networkExecutionState objects by then too.

Comment (yinghai) [reply lib/Runtime/HostManager/HostManager.cpp]: Yeah, it just doesn't necessarily need to be maxActiveRequests. One reason is that device resident tensors are going to consume resource in the cards. The other reason is that we can actually control the use of resource just by block waiting for the item to be available in the pool. Right now, let's put maxActiveRequests as an estimate. But we need to be aware that this number can change and system should still work.

File: lib/Runtime/Executor/CMakeLists.txt [hunk 5]
```diff
@@ -1,5 +1,6 @@
 add_library(Executor
               ExecutionState.cpp
```
Comment (yinghai) [inline lib/Runtime/Executor/CMakeLists.txt]: Where do we use ExecuteState anymore?

File: lib/Runtime/Executor/ThreadPoolExecutor.cpp [hunk 6]
```diff
@@ -207,6 +194,8 @@ void ThreadPoolExecutor::handleDeviceManagerResult(
       }
     }
   }
+  // Return intermediateContext to executionState.
+  executionState->returnUniqueNodeContextPtr(node, std::move(ctx));
```
Comment (yinghai) [inline lib/Runtime/Executor/ThreadPoolExecutor.cpp]: I found this context check in and check out a bit awkward. Maybe if we use raw point in the lower level calls can improve the situation. You don't need to change this now.
